src/library/scala/collection/immutable/Vector.scala

```diff
@@ -310,145 +353,158 @@ final class Vector[+A] private[immutable] (private[collection] val startIndex: I
+            val s = new Vector(startIndex - 1, endIndex, newBlockIndex)
             s.initFrom(this)
             s.dirty = dirty
-            s.shiftTopLevel(0, shiftBlocks) // shift right by n elements
-            s.gotoPosWritable(newFocus, newBlockIndex, newFocus ^ newBlockIndex) // prepare for writing
-            s.display0(shift - 1) = value.asInstanceOf[AnyRef]
+            s.gotoFreshPosWritable(newFocus, newBlockIndex, newFocus ^ newBlockIndex)
+            s.display0(lo) = value.asInstanceOf[AnyRef]
             s
           }
-        } else if (blockIndex < 0) {
-          // case B: we need to move the whole structure
-          val move = (1 << (5 * (depth + 1))) - (1 << (5 * depth))
-          val newBlockIndex = blockIndex + move
-          val newFocus = focus + move
-
-          val s = new Vector(startIndex - 1 + move, endIndex + move, newBlockIndex)
-          s.initFrom(this)
-          s.dirty = dirty
-          s.gotoFreshPosWritable(newFocus, newBlockIndex, newFocus ^ newBlockIndex) // could optimize: we know it will create a whole branch
-          s.display0(lo) = value.asInstanceOf[AnyRef]
-          s
-        } else {
-          val newBlockIndex = blockIndex
-          val newFocus = focus
-
-          val s = new Vector(startIndex - 1, endIndex, newBlockIndex)
-          s.initFrom(this)
-          s.dirty = dirty
-          s.gotoFreshPosWritable(newFocus, newBlockIndex, newFocus ^ newBlockIndex)
-          s.display0(lo) = value.asInstanceOf[AnyRef]
-          s
         }
-      }
-    } else {
-      // empty vector, just insert single element at the back
-      val elems = new Array[AnyRef](32)
-      elems(31) = value.asInstanceOf[AnyRef]
-      val s = new Vector(31, 32, 0)
-      s.depth = 1
-      s.display0 = elems
-      s
-    }
+      } else Vector.single(value)
+
     releaseFence()
     result
   }
 
   override def appended[B >: A](value: B): Vector[B] = {
-    val result = if (endIndex != startIndex) {
-      val blockIndex = endIndex & ~31 // round down to nearest 32
-      val lo = endIndex & 31 // remainder of blockIndex / 32
-
-      if (endIndex != blockIndex) {
-        val s = new Vector(startIndex, endIndex + 1, blockIndex)
-        s.initFrom(this)
-        s.dirty = dirty
-        s.gotoPosWritable(focus, blockIndex, focus ^ blockIndex)
-        s.display0(lo) = value.asInstanceOf[AnyRef]
+    val thisLength = length
+    val result =
+      if (depth == 1 && thisLength < 32) {
+        val s = new Vector(0, thisLength + 1, 0)
+        s.depth = 1
+        val newDisplay0 = new Array[AnyRef](thisLength + 1)
+        System.arraycopy(display0, startIndex, newDisplay0, 0, thisLength)
+        newDisplay0(thisLength) = value.asInstanceOf[AnyRef]
+        s.display0 = newDisplay0
         s
-      } else {
-        val shift = startIndex & ~((1 << (5 * (depth - 1))) - 1)
-        val shiftBlocks = startIndex >>> (5 * (depth - 1))
+      } else if (thisLength > 0) {
```

Comment:
Mostly indentation, changing condition to have a bit more obvious of logic:

old:
`if (startIndex != endIndex)`
new:
`if (thisLength > 0) // thisLength already computed anyways`